File: Stack, Queue and Linked Lists/stack_implementation.py
class Stack:
    # creating a new data type of class
    
    def __init__(self, size):
    # the constructor
        self.size = size
        # only the size will change here, the rest will remain the same
        # hence we only set size dynamically
        self.array = []
        self.top = -1

    def display(self):
        # this function prints the stack
        print(self.array[::-1])

    def get_top(self):
        # this gives the value to top at any given point in the program
        print("The value of top is",self.top)

    def push(self, value):
        # this function is used to add values to the stack
        if self.top < self.size - 1:
            self.array.append(value)
            self.top += 1
        else:
            print("The stack is full hence more elements cannot be added")

    def pop(self):
        # this function is responsible for removing values from the stack
        if self.top != -1:
            self.array.remove(self.array[-1])
            self.top -= 1
        else:
            print("The stack is empty hence no elements to remove")

if __name__ == "__main__":
    # making sure the code that runs below only runs in this file
    # in case the file wants to be used as a module
    # this makes it so that one can import the class and use it anywhere 
    Stack1 = Stack(5)
    # creating a stack object
    while(True):
        # Instructions on how to use the program
        checker = int(input("Type 1 to pop an element \nType 2 to push an element \nType 3 to display the array \nType 4 to display the value of top\nType 0 to stop the code\n"))
        if checker == 1:
            Stack1.pop()
        elif checker == 2:
            val = input("Enter the value to be pushed: ")
            Stack1.push(val)
        elif checker == 3:
            Stack1.display()
        elif checker == 4:
            Stack1.get_top()
        elif checker == 0:
            break

File: Stack, Queue and Linked Lists/StackImplementation.cpp
#include <iostream>
// we will be using vectors for this because vectors can dynamically
// allocate memory and can improve the overrall performance of the code
#include <vector>
using namespace std;


class Stack 
// creating a custom stack data type to add functionality 
{
    vector<int> array;
    // using a dynamic array to use as a stack
    int top;
    int size;

public:

    Stack(int size)
    // this is a constructor for the stack object
    {
        this->size = size;
        this->top = -1;
        this->array = {};
    }

    void push(int value); // this pushes value to the stack
    void pop(); // pops the last element from the stack
    void displayStack(); // displays the stack
    int getTop(); // helper function that returns the top of the stack
};

// function declarations are outside the class
void Stack :: push(int value)
{

    if (this->size != this->array.size())
    // checking if the stack is full or not 
    // .size() returns the size of the stack currently 
    {
        // .push_back() adds a value to the vector array
        this->array.push_back(value); 
        cout << "The value " << value << " has been pushed to the stack" << endl;
        this->top++;
        cout << "The value of top is " << getTop() << endl;
    } 
    else 
    {
        cout << "The stack is full" << endl;
    }
} 

int Stack :: getTop()
{
    // helper function
    return this->top;
}

void Stack :: pop()
{       
    if (this->size > this->array.size())
    // checking whether the stack is empty
    {
        cout << "The last element of the stack has been removed" << endl;
        this->array.erase(this->array.end() - 1); 
        // this line removes the last element of the vector
        // .erase() removes any element of a particular index
        // .end() - 1 gives the index of the last element of the stack
        this->top--;
        cout << "The value of top is " << getTop() << endl;
    } 
    else 
    {
        cout << "The stack is empty" << endl;
    }
}

void Stack :: displayStack()
{
    if (this->top != -1)
    // checking whether the stack is empty
    {
        cout << "The stack is: ";
        for (auto &element : this->array)
        {
            cout << element << " ";
        }
        cout << endl;
    }
    else 
    {
        cout << "The stack is empty." << endl;
    }
}



int main()
{
    // the driver code for the program
    int numOfElements;
    cout << "Enter the number of elements in the stack" << endl;
    cin >> numOfElements;
    Stack s1(numOfElements);
    // creating a stack object and dynamically giving it a size
    int choice;
    while (true)    
    //  entering a for loop
    {
        cout << "Enter a number from 0 to 4" << endl;
        cout << "0 : To exit program" << endl;
        cout << "1 : To push value" << endl;
        cout << "2 : To pop value" << endl;
        cout << "3 : To print the value of top" << endl;
        cout << "4 : To print the stack" << endl;
        cout << "Enter your choice" << endl;
        cin >> choice;
        switch (choice) 
        {
            case 1:
                int value;
                cout << "Enter a value to be pushed to the stack" << endl;
                cin >> value;
                s1.push(value);
                continue;
            case 2:
                s1.pop();
                continue;
            case 3:
                cout << "The value of top is: " << s1.getTop() << endl;
                continue;
            case 4:
                s1.displayStack();
                continue;
            case 0:
                cout << "Ending program" << endl;
                break;
            default:
                cout << "Enter a valid value" << endl;
                continue;
        }
        break;  
    }
    return 0;
}